### ISSUE TITLE
Database Document Fixtures

### ISSUE THREAD

**Issue by BlakeASmith** [virtool/virtool-workflow#82]
# Background

Each *workflow type* needs to access different documents from various collections in the database. In the current implementation, the id's for each document (analysis document, sample document, etc.) are given in the `job_args` (`job_document["args"]`). The job document itself is identified by the job's ID and is fetched before the workflow is started. The workflows then use the various ID's to request documents individually from the database. 

Presumably, if an ID for a document is provided in the `job_args` it is expected that the document will be used within the workflow. Thus it makes sense to lookup all the documents ahead of time and provide them to the workflow as fixtures.

Additionally, as a security measure, a workflow should **only** access those documents which are specified in their job document. As such it doesn't make sense to provide a general `lookup_document_by_id()` function. Instead, the workflow can be provided with all of the documents which it may need based on those assigned in the `job_args`.

# Plans To Remove Direct Database Access

For security reasons, it is planned to re-factor all existing workflows to avoid accessing the database directly. Idealy workflows should only be able to read and update the database in specific ways, and should be restrained to particular documents. 

Additionaly a move away from no-SQL MongoDB to an SQL database is also in the works. Thus the workflow code needs to be decoupled from `MongoDB` and provided with an interface which can remain the same once the database is migrated to an SQL solution. 

Eventually, all database access will be performed via an HTTP API or similar. Creating a module or abstract class where all database access is performed will be a good move in this direction. This will make moving the database calls to an external process via an HTTP API simpler. It will also simplify testing by allowing the database to be easily mocked, avoiding the need for a MongoDB process in many tests. 

# Fetching The Documents

The documents required for a workflow will be identified by the ID's listed in the `job_args`. For 
example, the `sample_id` field being present in the `job_args` would identify that a `sample_documet` fixture should be created. 

In abstract, the following should be done before the workflow is started. 

1. Use the `job_id` to lookup the `job_document` and get the `job_args`.
2. Identify ID fields in the `job_args` (sample_id, analysis_id, etc.).
3. For each ID found, fetch the document from the corresponding database collection. 
4. Create a fixture `*_document` for each document fetched. 

The workflow should be able to access the various documents as follows:

```python
from virtool_workflow import step

@step
def use_documents(sample_document, analysis_document, reference_document, ...):
	...
```

Where, for this workflow, the `job_args` would contain `"sample_id", "analysis_id", "reference_id"`.

# Updating The Documents In The Database

There must also be some mechanism provided for the workflow to update these documents, assuming they have the permission/right to do so. 

Currently the workflows just use the `Collection` object to modify the database directly, via the `update_one`, `delete_one` methods. Though a similar interface could be provided in the database module or abstract class, where the workflow developer provides the document ID and the changes to be made, it doesn't seem paricularly elegant in this case. 

Additionally we should try to avoid using mongo-specific features such as Mongo's update operators. If we continue to rely on mongo specific features instead of a more general API it will be a massive headache to translate those features into the SQL context later. Instead we should base the API around standard dictionary operations, which can be implemented separately in the SQL and noSQL contexts. 

As such, I think that bundling the update functionality into the fixtures themselves would be a good way to go. Instead of supplying a pure `dict`, the fixtures could provide a subclass of [MutableMapping](https://docs.python.org/3/library/collections.abc.html#collections.abc.MutableMapping) (I'll call it `VirtoolDocument`) which includes additonal methods for performing the database updates.

Within the workflow, the `VirtoolDocument` object would act exactly like a normal `dict`, able to be modified without updating the database in any way. When the workflow developer want's to change the underlying data in the database itself, they can use the additional `async` methods included in the `VirtoolDocument` class.

```python
from virtool_workflow import step, VirtoolDocument

@step
def update_document(some_document: VirtoolDocument):
	await some_document.set("field", "value")
	await some_document.remove("field")
	await some_document.create_new_field("name", "value")
```

## Use `dictdiffer` And a Single `sync()` Function

The [dictdiffer](https://github.com/inveniosoftware/dictdiffer) library can be used to discover the differences between two dictionaries. This can also be accomplished directly using set difference.  By comparing the differences between an old version of the document and the latest one, it could be determined which updates need to be made to the database document. 

This would allow for a simple API like this;

```python
@step
def update_document(some_document: VirtoolDocument):
	some_document["some_field"] = "some_value"
	...
	# Apply the local changes to the database document.
	await some_document.sync()
```

If only specific fields can be added/removed/modifed, we can raise an Exception or Warning when the workflow developer attempts to sync an illegal field. Alternatively, since we are using our own `MutableMapping` subclass, we could raise an Exception at the time the illegal field is added/removed/modified.

**Comment by igboyes** [comment]
Closing in favour of higher-level abstractions in fixtures (see #85 and #86).
